controllers/cloud.redhat.com/clowdapp_controller.go

```diff
@@ -138,192 +132,27 @@ func (r *ClowdAppReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
-		return ctrl.Result{}, nil
-	}
-
-	ans := &core.Namespace{}
-	if getAppNSErr := r.Client.Get(ctx, types.NamespacedName{Name: app.Namespace}, ans); getAppNSErr != nil {
-		return ctrl.Result{Requeue: true}, getAppNSErr
-	}
-
-	if ans.ObjectMeta.DeletionTimestamp != nil {
-		log.Info("App namespace is to be deleted - skipping reconcile")
-		return ctrl.Result{}, nil
-	}
-
-	if env.Generation != env.Status.Generation {
-		r.Recorder.Eventf(&app, "Warning", "ClowdEnvNotReconciled", "Clowder Environment [%s] is not reconciled", app.Spec.EnvName)
-		log.Info("Env not yet reconciled")
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, fmt.Errorf("clowd env not reconciled")); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
-		}
-		return ctrl.Result{}, nil
-	}
-
-	if !env.IsReady() {
-		r.Recorder.Eventf(&app, "Warning", "ClowdEnvNotReady", "Clowder Environment [%s] is not ready", app.Spec.EnvName)
-		log.Info("Env not yet ready")
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, fmt.Errorf("clowd env not ready")); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
-		}
-
-		return ctrl.Result{Requeue: true}, nil
-	}
-
-	cacheConfig := rc.NewCacheConfig(Scheme, errors.ClowdKey("log"), ProtectedGVKs, DebugOptions)
-
-	cache := rc.NewObjectCache(ctx, r.Client, cacheConfig)
-	//cache := providers.NewObjectCache(ctx, r.Client, scheme)
-
-	provider := providers.Provider{
-		Client: r.Client,
-		Ctx:    ctx,
-		Env:    &env,
-		Cache:  &cache,
-		Log:    log,
-	}
-
-	if provErr := r.runProviders(log, &provider, &app); provErr != nil {
-		r.Recorder.Eventf(&app, "Warning", "FailedReconciliation", "Clowdapp requeued [%s]", app.GetClowdName())
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, provErr); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
-		}
-		log.Info("Provider error", "err", provErr)
-		return ctrl.Result{Requeue: true}, provErr
-	}
-
-	cacheErr := cache.ApplyAll()
-
-	if cacheErr != nil {
-		r.Recorder.Eventf(&app, "Warning", "FailedReconciliation", "Clowdapp requeued [%s]", app.GetClowdName())
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, cacheErr); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
+	reconciliation := ClowdAppReconciliation{
+		ctx:      &ctx,
+		client:   r.Client,
+		recorder: r.Recorder,
+		app:      &app,
+		log:      &log,
+		req:      &req}
+	res, err := reconciliation.Reconcile()
+	if err != nil {
+		if shouldSkipReconciliation(err) {
```

Comment:
Sorry @adamrdrew I think there are some issues with this still. The main problem as far as I see is how the requeue is to be handled. In k8s, if you return a `nil` error from a reconciliation, the requeue request is ignored. It actually won't reconcile again ever, (until another requests comes in to reconcile). So this code returning `res` along with `nil` is actually equivalent to passing `ctrl.Result{}, nil`.

Reply:
Hmmm, is this a problem with the new implementation or what the problem always there?

This is the logic in the original code that I based this on:
```golang
	if ens.ObjectMeta.DeletionTimestamp != nil {
		log.Info("Env target namespace is to be deleted - skipping reconcile")
		return ctrl.Result{}, nil
	}
```

In our new step code that is recast as:
```golang
	if ens.ObjectMeta.DeletionTimestamp != nil {
		r.log.Info("Env target namespace is to be deleted - skipping reconcile")
		return ctrl.Result{}, errors.New(SKIPRECONCILE)
	}
```

Which when it is passed through the code you highlighted will result in returning `ctrl.Result{}, nil` which should be exactly the same thing we did in the original code.

The other place in the steps we do this is:
```golang
func (r *ClowdEnvironmentReconciliation) setToBeDisabled() (ctrl.Result, error) {
	if r.env.Spec.Disabled {
		r.log.Info("Reconciliation aborted - set to be disabled")
		return ctrl.Result{}, errors.New(SKIPRECONCILE)
	}
	return ctrl.Result{}, nil
}
```

Which was based on and should be logically equivilent to:
```golang
	if app.Spec.Disabled {
		log.Info("Reconciliation aborted - set to be disabled")
		return ctrl.Result{}, nil
	}
```

So, from the way I parse all this we should be doing exactly what the old implementation was doing. Am I parsing this out wrong? If not, then is this a problem that already existed in Clowder? And if so, what so I do in these cases?

Reply:
```
func (r *ClowdAppReconciliation) isEnvLocked() (ctrl.Result, error) {
	if ReadEnv() == r.app.Spec.EnvName {
		r.recorder.Eventf(r.app, "Warning", "ClowdEnvLocked", "Clowder Environment [%s] is locked", r.app.Spec.EnvName)
		r.log.Info("Env currently being reconciled")
		return ctrl.Result{Requeue: true}, errors.New(SKIPRECONCILE)
	}
	return ctrl.Result{}, nil
}
```

I think this function is the one that breaks. It says the Clowder Env is Locked, and it should be requeued. But it we squash the error and return `nil`, this won't be the case.

Reply:
OK, I changed it but that's what we did before this change:

```golang
	if ReadEnv() == app.Spec.EnvName {
		r.Recorder.Eventf(&app, "Warning", "ClowdEnvLocked", "Clowder Environment [%s] is locked", app.Spec.EnvName)
		log.Info("Env currently being reconciled")
		return ctrl.Result{Requeue: true}, nil
	}
}
```

Reply:
We're failing on merge conflict again. There were more changes to the deployment implementation so we're dying on that. I tried to resolve it but I'm not sure what changed - the diff output is a mess again due to how big the delta is in this branch.

Reply:
I may have a large apology to make

Reply:
I think I had forgotten the state we had where this was wrong
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler

Reply:
Cool cool - flipped it back around